### ISSUE TITLE
Ipadapter problem after commit c28... (comfyui won't start generating)

### ISSUE THREAD

**Issue by patientx** [cubiq/ComfyUI_IPAdapter_plus#109]
PC : windows 10, 16 gb ddr4-3000, rx 6600, using directml with no additional command parameters. 

Updated today with manager , and tried my usual workflow which has ipadapter included for faces, when it comes to actually generating the image , it just stops there, nothing happens. I have to close the command prompt and relaunch the app.  

Loading 1 new model
  0%|                                                                                                                                                           | 0/8 [00:00<?, ?it/s]

I was mainly using lcm so tried normal sd1.5 models , the same results. Tried different samplers , same.
If I get ipadapter out of the workflow everything works correctly. So I am not updating daily , thought something in the middle changed, tried every commit and found out that it works up until "[timestepping,](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/1ed4f93d1b757fd626d255e734f8be18af5b72a7) [fix](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/1ed4f93d1b757fd626d255e734f8be18af5b72a7)" , after that the problem is here. 
So now I am on c28a04466b17d760a345aea41d6a593c0a312c95. (last one before that fix) . Everything works as it was before.

**Comment by cubiq** [comment]
@laksjdjf it's your code I believe, can you take a look at it?

We can revert that change without issues.

@patientx can you post your startup message from comfyui?

**Comment by cubiq** [comment]
sorry I misunderstood... so it's [1ed4f93](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/1ed4f93d1b757fd626d255e734f8be18af5b72a7) that is not working.

did you update comfyui?

**Comment by patientx** [comment]
I was use comfyui latest version yes. Btw at first I tried using previous commits of comfyui and it was around 30 commits before that the extension at latest version worked, so I thought comfy is the main app and the latest additions are more important if I can fix the problem with the node. So here we are :)

Here is comfyui startup ,
https://pastebin.com/a6AUGPNu

For the record I tried deleting all the other custom nodes except this one, still had the same problem.

**Comment by patientx** [comment]
comfyui updated and they say they resolved the ipadapter issue, so I updated both comfy and this node to the latest, nope still the same issue.

**Comment by cubiq** [comment]
the latest version overrides timestepping if it's not supported, so I really don't know how to help. If it doesn't give any error I would tend to think of some drivers issue maybe? Or maybe you need to upgrade also the python libraries. Can you try with the latest nightly (instead of the stable one?)

**Comment by MGTRIDER** [comment]
Hi there i can vouch that this is not a driver problem as today i was still using your nodes fine, until i updated the nodes including comfyui to the latest version. What basically happens is that cmd windows shows that inference is starting but then it literally keeps sitting there on the first step until after a while comfyui just crashes.

**Comment by MGTRIDER** [comment]
If it helps, i am on this release: ComfyUI Revision: 1758 [6b769bca] | Released on '2023-11-30'. And i am also using an AMD gpu. As i said before everything was still working fine today until i updated to the latest commit.

**Comment by cubiq** [comment]
I feel this is more of a comfyui issue than IPAdapter, but I'll check with comfyanonimous

do you get any error at all?

**Comment by MGTRIDER** [comment]
No, 

> I feel this is more of a comfyui issue than IPAdapter, but I'll check with comfyanonimous
> 
> do you get any error at all?

No, no error. Everything loads fine until, the part where it starts to generate the image, then it it just hangs on step 0 of 40. But as soon as i bypass the ip adapter node, everything works fine again. And it also doesn't matter if i use time stepping or not, neither if batch unfold is on or off. It does the same thing.

**Comment by cubiq** [comment]
if you do not use timestepping or batch unfold the code in unchanged. It's just an IF statement so if you don't use either everything is the same as before.

**It is important that you refresh the browser cache** (or use an incognito window), delete the IPAdapter nodes and recreate them.

If the execution halts and then crashes without errors it is very likely a driver issue.

I talked to Comfyanonymous this morning that confirmed my suspects:

![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/427614/f8207c1a-a092-4616-adb7-54f8aa7cb632)

Unfortunately DirectML is very unstable on Windows and it would require a lot of debugging to reach the source of the problem and it could be even related to a specific GPU.

Anyway I made a small change just now that could help (don't hold your breath though). You can give it a try.

**Comment by MGTRIDER** [comment]
I deleted the node manually, used git clone to copy the repository with the change you made. Sadly no dice, still the same thing. One question, is it normal that in cpu mode it also doesn't work ? Seems like AMD users are getting the short end of the stick again. But even so, thank you for this great node you made. It was fun to work with. Keep doing the good work. Thumbs up to you sir.

**Comment by cubiq** [comment]
in CPU mode could be just some kind of overflow, I'll see if I can replicate

**Comment by MGTRIDER** [comment]
In cpu mode when i use the standard quad-cross-attention, it hangs the the same as in directml mode. When i try to use split-cross-attention, it will complain about a 'cuda' or 'cpu' error unless i force unet-bf16. But then it hangs the same again lol.

**Comment by patientx** [comment]
just use the old commit , everything is working there

**Comment by MGTRIDER** [comment]
> just use the old commit , everything is working there

Hi there, do you mean the commit of ipadapter plus before the timestepping feature ?
If so, how do i go back to the version before the commit ?

**Comment by MGTRIDER** [comment]
Never mind, i found out how to do it. Thanks for the suggestion. It's still weird though how this commit works fine but since the timestepping commit inference just refuses to work. More so when i use a controlnet node tthat has timestepping feature that works pretty much fine.

**Comment by cubiq** [comment]
does 1ed4f93 work?

**Comment by cubiq** [comment]
if you are on c28a044 can you place this code at line 388 and tell me what it says? (delete them afterwards)

```
        print(model.model.model_sampling.percent_to_sigma(0.0))
        print(model.model.model_sampling.percent_to_sigma(1.0))
```

Thanks
(sorry I don't know how else to test this)

**Comment by MGTRIDER** [comment]
> if you are on [c28a044](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/c28a04466b17d760a345aea41d6a593c0a312c95) can you place this code at line 388 and tell me what it says in the command prompt? (delete them afterwards)
> 
> ```
>         print(model.model.model_sampling.percent_to_sigma(0.0))
>         print(model.model.model_sampling.percent_to_sigma(1.0))
> ```
> 
> Thanks (sorry I don't know how else to test this)

Hi there, in which file should i paste the codes ?

**Comment by cubiq** [comment]
IPAdapterPlus.py

please note you need to keep the indentation! of the original file

**Comment by MGTRIDER** [comment]
So these are the two codes i see at line 388:

def apply_ipadapter(self, ipadapter, model, weight, clip_vision=None, image=None, weight_type="original", noise=None, embeds=None, attn_mask=None, start_at=0.0, end_at=1.0):

Should i overwrite them ?

**Comment by cubiq** [comment]
I don't know what version you are on... just put them above the line

`self.dtype = model.model.diffusion_model.dtype`

at the same indentation level, do no remove anything.

**Comment by MGTRIDER** [comment]
> I don't know what version you are on... just put them above the line
> 
> `self.dtype = model.model.diffusion_model.dtype`
> 
> at the same indentation level, do no remove anything.

I am testing commit c28a044 that doesn't work.

**Comment by cubiq** [comment]
you can put them in any commit you are on, it doesn't matter. I just need to know what values are returned. You should see two long float numbers in the command window

**Comment by MGTRIDER** [comment]
![Screenshot 2023-12-02 054939](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/106754440/1b34971d-e6d2-41ba-8df5-ec1ca6051644)

**Comment by MGTRIDER** [comment]
Should i also do one for the commit that doesn't work ?

**Comment by cubiq** [comment]
no, that's fine. thanks. the numbers are correct, so I'm really out of ideas here.

**Comment by MGTRIDER** [comment]
> no, that's fine. thanks. the numbers are correct, so I'm really out of ideas here.

Haha, i understand. Even so thanks again. At least i can still use your node. Even if i have to stay on an older version for now. It gets the job done. So again, thumbs up to you and keep doing good work.

**Comment by cubiq** [comment]
unless anyone is willing to give me access to their PC remotely...

**Comment by MGTRIDER** [comment]
I'd like to but i can't really risk it as i also use my PC for work. Not just for hobbies. So i hope you understand.

**Comment by cubiq** [comment]
no worries, I wouldn't trust a rando on my own PC either

**Comment by OtakuD** [comment]
In case its helpful or related I get this error when using on AMD too:
ComfyUI\comfy\model_sampling.py:74: UserWarning: The operator 'aten::frac.out' is not currently supported on the DML backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at D:\a\_work\1\s\pytorch-directml-plugin\torch_directml\csrc\dml\dml_cpu_fallback.cpp:17.)
  w = t.frac()

**Comment by MythicalChu** [comment]
Echo'ing the "I'm on AMD, same issue right after updating IpAdapter_plus, bypassing it works fine"

**Comment by lord-lethris** [comment]
roling back to [c28a044](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/c28a04466b17d760a345aea41d6a593c0a312c95) temporeroly fixed it for me.

Indecently - AMD are getting ready to RCom Windows (just saw it on AMDs homepage) - So hopefully no more DirectML restrictions on windows platform and AMD based AI should fly like nVidia..!

**Comment by patientx** [comment]
Here is weird one for you guys :  

I updated all extensions as usual with manager and before rolling back to c28.. I decided to try once again, behold it worked with the workflow I have at the time. 

And then I tried to build a default workflow with a clean sheet and .. it didn't work ??? What I had with the other workflow was I was experimenting with kohya's hires fix AND hypertile (to be able to generate images at higher quality somehow faster) 

Weird thing 1: I can't generate bigger than 512x960 (I normally generate at 512x768 or 512x512, with kohya I am able to twice that very easily but with this workflow it doesn't generate when I enter for example 768x960 ,512x1024 , or 1024x1536 (two times) LIKE in a1111 it is like my memory isn't enough to generate higher than that)

Weird thing 2: BOTH have to be used , if I disable one , the usual happens, no generation at start.

Weird thing 3: Kohya doesn't seem be working at all because there is no speed change in generation, but hypertile is definetly working if I change the tile size speed and end quality changes , (128 and low lowers the quality)

So this combo somehow lets me use ipadapter again. 

Maybe someone more informed could get something out of this situation for solving the problem.

[weird workload that enables ipadapter on amd.json](https://github.com/cubiq/ComfyUI_IPAdapter_plus/files/13662754/weird.workload.that.enables.ipadapter.on.amd.json)

**Comment by jtary** [comment]
> rolling back to [c28a044](https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/c28a04466b17d760a345aea41d6a593c0a312c95) temporarily fixed it for me.
> 
> Indecently - AMD are getting ready to ROCm Windows (just saw it on AMDs homepage) - So hopefully no more DirectML restrictions on windows platform and AMD based AI should fly like nVidia..!

Unfortunately MIOpen (which is the part of ROCm that pyTorch depends on) is further behind than ROCm in general. There is active work going on, but I think we'll be lucky to get a ROCm enable windows pytorch before 2025.

**Comment by cubiq** [comment]
can any of you try the latest update?

**Comment by patientx** [comment]
> can any of you try the latest update?

This error comes now : 
`Requested to load CLIPVisionModelProjection
Loading 1 new model
ERROR:root:!!! Exception during processing !!!
ERROR:root:Traceback (most recent call last):
  File "D:\ComfyUI\execution.py", line 153, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "D:\ComfyUI\execution.py", line 83, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "D:\ComfyUI\execution.py", line 76, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "D:\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 464, in apply_ipadapter
    image_prompt_embeds, uncond_image_prompt_embeds = self.ipadapter.get_image_embeds(clip_embed.to(self.device, dtype=self.embeds_dtype), clip_embed_zeroed.to(self.device, dtype=self.embeds_dtype))
  File "D:\ComfyUI\venv\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "D:\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 200, in get_image_embeds
    image_prompt_embeds = self.image_proj_model(clip_embed)
  File "D:\ComfyUI\venv\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "D:\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 41, in forward
    clip_extra_context_tokens = self.proj(image_embeds)
  File "D:\ComfyUI\venv\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "D:\ComfyUI\venv\lib\site-packages\torch\nn\modules\container.py", line 217, in forward
    input = module(input)
  File "D:\ComfyUI\venv\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "D:\ComfyUI\venv\lib\site-packages\torch\nn\modules\linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: mat1 and mat2 must have the same dtype`

Even with newly built,  just basic workflow. Previous versions at least worked with the trick I mentioned above. (if I use kohya hires fix + hypertile ; ipadapter works. I explained it a few posts up)

**Comment by cubiq** [comment]
what about now?

**Comment by patientx** [comment]
> what about now?

That mat1 .. error disappeared , same problem as before, but still working with my method at least

**Comment by jtary** [comment]
Just gave it a try with https://github.com/cubiq/ComfyUI_IPAdapter_plus/commit/cb1c5e8a1af794a605092da78218ddaa06615235, seeing the same issue where the IP Adapter step runs fine, but when it gets to the sampler it just get stuck at 0%.

```
Requested to load BaseModel
Loading 1 new model
  0%|                                                                                           | 0/30 [00:00<?, ?it/s]
```

**Comment by JarekDerp** [comment]
I run some tests, edited the py script and I came to one conclusion - directml doesn't like the "sigma_start" and "sigma_end".
Once I removed them from the script (latest commit was 2023-12-17) it started to work again:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/2073634e-dec4-456f-9f50-07b521b59ca0)

Obviously when I change start_at and end_at in the webui then it doesn't do anything.

**Comment by patientx** [comment]
> I run some tests, edited the py script and I came to one conclusion - directml doesn't like the "sigma_start" and "sigma_end". Once I removed them from the script (latest commit was 2023-12-17) it started to work again: ![image](https://private-user-images.githubusercontent.com/54577977/291612618-2073634e-dec4-456f-9f50-07b521b59ca0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5OTczMDMsIm5iZiI6MTcwMjk5NzAwMywicGF0aCI6Ii81NDU3Nzk3Ny8yOTE2MTI2MTgtMjA3MzYzNGUtZGVjNC00NTZmLTlmNTAtMDdiNTIxYjU5Y2EwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDE0NDMyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1ZjczODRjNzhmZmFhMTc5YTE0OTU0N2YxYjY5MjVkN2QzZGQwYWZhNGM1Zjg4MWMzMzU3MGEzOTYyMGIyMTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eC8VWNqcxEkCYNf_6JTerriiew2e3aQrGWEf793e73M)
> 
> Obviously when I change start_at and end_at in the webui then it doesn't do anything.

Like all of them ? Can you share the edited file ?

**Comment by cubiq** [comment]
> I run some tests, edited the py script and I came to one conclusion - directml doesn't like the "sigma_start" and "sigma_end". Once I removed them from the script (latest commit was 2023-12-17) it started to work again:

without the code this doesn't help much

**Comment by JarekDerp** [comment]
Okey, I pin-pointed it to two lines to edit. It's these two.
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/64df5600-9aaf-4be5-acc9-7e7e6f9d2619)

After removing them like in the screenshot then it works fine. The "start_at" and "end_at" parameters gets ignored but it generates the image and I assume it applies it at the whole generation from 0 to 100%.

**Comment by cubiq** [comment]
mh I believe the problem is the hardcoded value.

please print(extra_options["sigmas"][0].item()) at line 236
then before line 250 print(sigma_start, sigma_end)

**Comment by cubiq** [comment]
just print(extra_options["sigmas"])

**Comment by JarekDerp** [comment]
I printed whole extra_options and here's what I got:
```
extra_options
{'cond_or_uncond': [0], 'sigmas': tensor([14.6146], device='privateuseone:0'), 'original_shape': [1, 4, 96, 64], 'transformer_index': 0, 'block': ('input', 1), 'block_index': 0, 'n_heads': 8, 'dim_head': 40}
999999999.9 0.0
```

**Comment by JarekDerp** [comment]
I made edits like this:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/0a1503fc-10ed-4f4f-8a6a-e6833eb96d57)


and it didn't work. It was stopping at line 237 without any explanation or error messages. Tried to remove the [0] and it didn't work either.
When I changed it to:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/376687d2-b3f5-433c-bc4e-9b9cb2a6e8b4)

It started to spit out some numbers:
```extra_options[sigmas].item()
tensor([14.6146], device='privateuseone:0')
if 'sigmas' in extra_options
Sigma in extra_options
999999999.9 0.0```
Very strange problem.

**Comment by cubiq** [comment]
if you can join Discord we can try to fix it (I would be available in about 1:30h)

**Comment by JarekDerp** [comment]
Where can I find the discord invite link?

I feel like I'm getting closed. If only I knew what I'm doing.
I changed the code to something like this:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/18f3bc70-737c-404c-8006-26a970d79ed6)

And now it sometimes work and sometimes not. On the preview in KSampler I can see that it sometimes gets to step 5, but then starts to show a black image. Sometimes it even manages to get to step 8, but most often I get a black image from the beginning.
[log.txt](https://github.com/cubiq/ComfyUI_IPAdapter_plus/files/13717996/log.txt)

**Comment by cubiq** [comment]
https://latent.vision/discord

**Comment by JarekDerp** [comment]
Ok, I got it to work. I don't know why, but it works and it even respects the start_at and end_at parameters correctly. The black image was just some issue with my ComfyUI. Looks like the problem was some strange behaviour when trying to do extra_options["sigmas"][0].item(). When I split it into two part it works fine.

```
         org_dtype = n.dtype
         cond_or_uncond = extra_options["cond_or_uncond"]
-        sigma = extra_options["sigmas"][0].item() if 'sigmas' in extra_options else 999999999.9
-
+        temp_sigma = extra_options["sigmas"] if 'sigmas' in extra_options else null
+        sigma = temp_sigma[0].item() if temp_sigma else 999999999.9
+        
         # extra options for AnimateDiff
         ad_params = extra_options['ad_params'] if "ad_params" in extra_options else None
```

**Comment by cubiq** [comment]
I would suggest to try this

```
        sigma = extra_options["sigmas"] if 'sigmas' in extra_options else None
        sigma = sigma[0].item() if sigma is not None else 999999999.9
```

**Comment by JarekDerp** [comment]
> I would suggest to try this
> 
> ```
>         sigma = extra_options["sigmas"] if 'sigmas' in extra_options else None
>         sigma = sigma[0].item() if sigma is not None else 999999999.9
> ```

This doesn't work, unfortunately. The same as before, just freezes when it gets to KSampler.

**Comment by cubiq** [comment]
doesn't make any sense

**Comment by JarekDerp** [comment]
I know. It doesn't make sense at all, even though I'm not familiar with Python.
Well, at least this one works for some reason:
```
        PointlessVariable = extra_options["sigmas"][0]
        sigma = PointlessVariable.item() if PointlessVariable else 999999999.9
```

Or this one works too:

```
        sigma = extra_options["sigmas"][0]
        sigma = sigma.item() if sigma else 999999999.9
```

As if it likes bad syntax.
When I was writing scripts in PowerShell I remember I used to have an issue where a command in line 89 got executed before the command in line 88. So for example code like this:
```
88  print("ABC")
89  print("abc")
```
got executed like this in the log:
```
abc
ABC
```

I get a similar vibe here.

**Comment by JarekDerp** [comment]
I think this needs to work like I'm the example here:
https://pytorch.org/docs/stable/generated/torch.Tensor.item.html
In the example they assigned a tensor to a variable, and then used the ".item()".

I think it works in a way so that it moves the information from GPU to CPU to be used in the calculation later and it cannot be used from GPU directly when using Directml.
That's why it works when assigned to a variable first.
That would also explain why it works fine when I run Comfyui on my CPU.

**Comment by cubiq** [comment]
```
sigma_fuuu_directml = extra_options["sigmas"] if 'sigmas' in extra_options else None
sigma = sigma_fuuu_directml[0].item() if sigma_fuuu_directml is not None else 999999999.9
```
so would this work?

**Comment by JarekDerp** [comment]
Nope, freezes. I tried many different things. I even tried "try except" like this:
```
        try:
            x = extra_options["sigmas"][0]
            sigma = x.item()
        except:
            sigma = 999999999.9
        print(sigma)
```
And nothing, still freezes while trying to read extra_options["sigmas"][0].item(). Looks like this is a problem with DirectML like comfyanonymous said.
If I do just extra_options["sigmas"][0], then the comfy doesn't freeze and it continues, with black image, but at least it continues. And freezes as soon as it hits ".item()".

Here's example. Running this:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/dd8aa1d9-7d96-439e-acc1-6e4d6adcdd04)

I got info like this and it froze:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/bdcc354b-d2f1-4f72-8ed7-5f5b73ef26fa)

I don't know why it didn't freeze in my previous examples. I tried looking into ttn nodes and efficiency nodes looking into how they've done it but I don't understand anything. I'm not good with Python as you might have guessed.

**Comment by jtary** [comment]
There seems to be some sort of race-condition going on. I can get the adapter to run, just by adding some prints:

```
+        print("Getting ready!")
+       print(f"Okay, about to get the item for {extra_options['sigmas']}")
         sigma = extra_options["sigmas"][0].item() if 'sigmas' in extra_options else 999999999.9
+        print(f"sigma: {sigma}")
```

I think `extra_options['sigmas']` needs to be accessed to set some internal state. Without that state, trying to reference `[0]` causes the call the block. Some kind of lazy evaluation maybe?

**Comment by JarekDerp** [comment]
I also noticed that adding some prints unblocks it but if I unblock it with prints then I get black images and comfy complains about some img decoding script problems (error message).
Oh well, I give up. I'm really considering just abandoning image generation altogether until AMD Devs finally starts supporting ROCm on Windows. One workaround after another just to generate a handful of decent images.

**Comment by jtary** [comment]
Yeah, it's frustrating. I really don't want to support nVidia, but both AMD and Microsoft seem to be really dropping the ball on supporting their platforms. `torch-directml` is woefully out of date, and while ROCm with Pytorch on Windows is moving forward, I'm doubtful it will have support for older hardware.

**Comment by JarekDerp** [comment]
For me the black image starts from step 1, but sometimes starts in the middle or even finishes without any problems.
I tried Microsoft Olive (onnx) recently and it works quite well. I get about 400%-500% better speeds on the models I optimized - the same speed as normal safetensor version of a given model when running AMD+Linux.
I could generate 1024*1024 image in less than 7 seconds on both Olive and Linux on my 6700XT.
But unfortunately there's a lot of things you cannot do with Olive at the moment and I hate using Linux.

**Comment by jtary** [comment]
Ok, seems to be another "you must access this variable before using" bug. You can work around it using this patch to comfy:

```
--- a/comfy/ldm/modules/attention.py
+++ b/comfy/ldm/modules/attention.py
@@ -523,6 +523,10 @@ class BasicTransformerBlock(nn.Module):
                 if value_attn2 is None:
                     value_attn2 = context_attn2
                 n = self.attn2.to_q(n)
+
+                if n.isnan().any():
+                    raise ValueError("n is nan")
+
                 context_attn2 = self.attn2.to_k(context_attn2)
                 value_attn2 = self.attn2.to_v(value_attn2)
                 n = attn2_replace_patch[block_attn2](n, context_attn2, value_attn2, extra_options)
```

**Comment by cubiq** [comment]
this happily falls in "won't fix". DirectML is a joke. If you hate nvidia and want to do ML with AMD, install Linux. It's harsh but true.

**Comment by jtary** [comment]
Understandable. I can't think of a more 'hacky' fix, than doing an error check to force a lazy-evaluation to resolve because there is a bug that was probably fixed 6 months ago, but direct-ml can't get their shit together. I guess if any other windows/amd folks come around asking about it, this issue serves as documentation of how you can hack a fix in yourself. 

Thanks @cubiq for trying to troubleshoot it.

**Comment by cubiq** [comment]
I could bypass the whole timestepping if I detect dml but I have no way to check if it will ever work, so I dunno. Need to think how to do it properly.

**Comment by jtary** [comment]
Even if you fix your issue, there is still the black-box NaN issue from Comfy. The real bug is in Torch, and was probably fixed 10 releases ago. Microsoft really needs to release a new version of torch-directml, so that's compatible with the most recent Torch versions. It's not like they don't have the resources to throw a few developers at it.

**Comment by patientx** [comment]
I understand the decision it is entirely on ms and amd at this point for various problems regarding directml (not only this) but would there be any problems (with the extension or comfyui in any other way) if I continue to use JarekDep's temp solution here  , working for me with the latest version this way.  https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/109#issuecomment-1863077024 . Thanks for actually trying to solve this. Hope amd gets their sh*t together.

Keep up the good work !

**Comment by lord-lethris** [comment]
I had this very problem today after updating ComfyUI and my Nodes.

The Temp Solution in https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/109#issuecomment-1862932994 also fixed it for me.

**Comment by aa022** [comment]
had the same problem, https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/109#issuecomment-1862932994 seems to work, tysm guys

**Comment by JarekDerp** [comment]
It's a shame we're not able to fix it for Directml. Over the past few days I've played around with the timestepping running it on the CPU and it's so good for the image quality. You have so much more control regarding how the image will look like. It would be great if someone's good with programming, Directml know-how and AMD card would have a look at it.

**Comment by JarekDerp** [comment]
Hi @cubiq
I think this time stepping doesn't work for CPU either.
I just set up "start_at = 0.5" and "end_at = 1" and it's still does it for all the steps.
I've only changed this:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/581c3712-43f4-4f14-a68f-ed138822fcbf)

Here's at step 1/20:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/a0e98931-cdcb-421e-bd23-dfe17afcab73)

Here's step 2/20:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/1e4a61d1-d422-4a1e-97d9-81090ad1fc12)

and so on until the end.
So if I'm running 9 steps in LCM, and I want to apply it to the last 5 steps only, then I would have to change "start_at" to around 0.80 for it to skip the first 4 steps. I tried different samplers (I assumed they might be calculating sigmas differently?) but it all looks the same. Only changing it to 0.80 make it apply to the ~50% of the steps.

I'm at the latest commit 4e898fe.

**Comment by gkiryaziev** [comment]
You can make these changes in the code, it won't change the logic, but it will solve the problem with AMD GPU and DirectML.

![2024-02-02_214400](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/93bc09be-76ab-49a4-95ca-ad3cbee734cf)

**Comment by cubiq** [comment]
does it work now?

**Comment by gkiryaziev** [comment]
**No, it doesn't work. It's stuck at 0%**


![1](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/e530b5cc-f587-4f8a-b737-b3ec679ba0f0)
![2](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/746aa992-8a22-4a0c-ba61-7c342468233b)


**But this is how it works. I removed "is not None" in second line**


![3](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/8f0f73e1-79e4-408f-b493-da4584e00c40)
![4](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/dd9514be-179a-4409-a91f-818eca98216a)
![5](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/68f090c6-0e14-472d-89ca-3bab7f23bd46)


**Here's the result**


![6](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/3f654439-fffa-443a-8e5c-9821f2a9b9c4)


**Perhaps the error is that the variable is of type float and doesn't pass the None check?**

**Comment by cubiq** [comment]
doesn't make any sense but if that's what it takes...

**Comment by cubiq** [comment]
actually I have to check if it may erroneously validate with just if sigma

**Comment by lord-lethris** [comment]
> You can make these changes in the code, it won't change the logic, but it will solve the problem with AMD GPU and DirectML.
> 
> ![2024-02-02_214400](https://private-user-images.githubusercontent.com/8873218/302035869-93bc09be-76ab-49a4-95ca-ad3cbee734cf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4NjEzNTUsIm5iZiI6MTcwNzg2MTA1NSwicGF0aCI6Ii84ODczMjE4LzMwMjAzNTg2OS05M2JjMDliZS03NmFiLTQ5YTQtOTVjYS1hZDNjYmVlNzM0Y2YucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTNUMjE1MDU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjA3Nzc4MTM2ZjE2ZGE4N2U5MGZjZDYzZDg3MTM2NTVhOGNiY2NlOGUxODc3ZDgxNzExYzcwOTVjZjg2MDk3MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QLFPoHCzTLVN7iTS4c9ULye6wtm1d6Z4L06uDk-lJrA)

This also fixed it for me after another update.

**Comment by gkiryaziev** [comment]
> doesn't make any sense but if that's what it takes...

The difference is that in this variant:

> sigma = extra_options["sigmas"][0] if 'sigmas' in extra_options else None
> sigma = sigma.item() if sigma else 999999999.9

sigma is assigned the value 999999999.9 if sigma evaluates to False. This means that if sigma is None, an empty list, or any other value that evaluates to False in a boolean context, it will be replaced by 999999999.9.

But in this variant:

> sigma = extra_options["sigmas"][0] if 'sigmas' in extra_options else None
> sigma = sigma.item() if sigma is not None else 999999999.9

sigma is assigned the value 999999999.9 only if sigma is None. If sigma is anything other than None, it will proceed to check if it has an .item() method and execute accordingly.

In the first line sigma has type torch.Tensor, and torch.Tensor is a multi-dimensional matrix. Maybe this object doesn't pass the None check, but it passes the False check.

**Comment by JarekDerp** [comment]
What exactly is this sigma? And in what instances it's "None" that we need to check?
Isn't it just information about a tensor's state that's currently working on diffusing the image in the current interference step? I would assume that it's always there, and the more steps it already finished, the closer it gets to value 0 (0 = no interference left in the image) - at least that's how I understand it.

So by doing
``
sigma = extra_options["sigmas"][0] if 'sigmas' in extra_options else None
sigma = sigma.item() if sigma else 999999999.9
``
First you are checking if there's a property "Sigmas" in extra_options and load the first instance of the sigma. If there's no Sigmas, then you assign it value "None".
Then second step just checks if it's "None" (from the previous if statement) or not ("None" or 0 = false; any other value will return true).

In summary, those two ifs has those outcomes:
No Sigma = assign 9999999.9 (telling the diffusor that it's far from finished and needs to diffuse the image)
Sigma exists = assign the value of the first sigma found

Isn't that right?

**Comment by JarekDerp** [comment]
Ah, I just realized that there's a potential problem with it as I was re-reading my message.
As sigma approaches 0 in the last steps, it might be that the last couple steps will have sigma value of 0.888 or 0.321 or something like that. I'm not sure if python will round it down to 0 and on the "if sigmas" it will return "False" and apply the 99999.9 - meaning the "end_at" value might be ignored for the last couple steps. I'm not sure how that works in python, but in Excel and PowerShell it works fine, but I don't have python to test it:
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/54d2d42f-5fa9-45ae-8a03-750a0b5b8c2a)
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/54577977/e1ee301c-0807-49d3-ba81-cef27cdd00a4)

**Comment by gkiryaziev** [comment]
So, I've added a few prints to your previous code, to the current code and to my own, without **is not None**.

![9](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/c91b30f9-8e41-475c-bf2e-b2df85280254)
![10](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/a3334a7e-e5a5-4785-9e9e-4da6ae0a86f2)

![11](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/e91f4d1c-8439-4491-878b-476caa3eabf8)
![12](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/02dc5c25-df67-45f4-a62e-563b04992e2d)

As we can see the **__call__** function is called iteratively and stuck on the second iteration in both cases.
But without **is not None** the code works and reaches the end.

![13](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/72554bbf-ac29-4ca4-b4cb-bf87fbb202d5)
![14](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/0ebce29e-f050-44d5-b5e7-6f49cc84ac3e)

Maybe this will help to understand the problem.

**Comment by JarekDerp** [comment]
@gkiryaziev we are not worried whenever it works or not. We know it works. I confirmed it in this comment https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/109#issuecomment-1863300846
What we are worrying about is whenever this change will break something. We are wondering "Should we kick this ball? It might break the window if I keep kicking it and it goes wrong" and you're here saying "Kick this ball into a wall. I kicked it once and it's fine.".

**Comment by cubiq** [comment]
does it work with `if isinstance(sigma, torch.Tensor)`?

**Comment by gkiryaziev** [comment]
**Unfortunately, that doesn't work either. Still stuck at 0%** 

![1](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/748c4644-9a10-410d-98f2-cec54374dbc7)

**Comment by gkiryaziev** [comment]
I think this is not a type checking issue, I think this is a timing issue, a delay in device initialization, because if I specify print exactly after first line, process goes on, doesn't get stuck at 0%, but result is a black image.


![1](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/dd1cfc83-5a09-4a65-acee-a48632c3cb99)
![2](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/cc5bb291-718c-4287-850f-1ce3bcf1d811)
![3](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/77af2b70-4e20-4c95-bbae-c628f98bc3a7)
![4](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/8873218/156f35b4-84f6-481d-82c4-08f4ed3a47f6)


Very strange problem, because torch.Tensor should pass the None check. And why the program changes its behavior just after adding one print operator.

**Comment by sanchitwadehra** [comment]
ok so currently if anyone is facing this issue still, here is the solution thanks to (JerekDerp) :- 

Problem: The original code attempted to access the first element of extra_options["sigmas"] and convert it to a Python number in a single line. However, if extra_options["sigmas"] did not exist or was None, this would raise an error because you can’t access an element of None.

Solution: The solution was to split this operation into two parts. First, we check if extra_options["sigmas"] exists and is not None. If it is, we set sigma to 999999999.9. If extra_options["sigmas"] does exist, we then access the first element and convert it to a Python number. This prevents any attempts to access an element of None, which would raise an error.

Here’s the modified part of the code:

Python Code change in IPAdapterPlus.py file at the :- 

# Split the operation into two parts
temp_sigma = extra_options["sigmas"] if 'sigmas' in extra_options else None
sigma = temp_sigma[0].item() if temp_sigma else 999999999.9


This modification ensures that sigma is always assigned a value, either from extra_options["sigmas"] or the default value 999999999.9, without raising an error. This makes the code more robust and prevents it from crashing due to a NoneType error when extra_options["sigmas"] does not exist or is None.

**Comment by cubiq** [comment]
can someone try `sigma = extra_options["sigmas"].cpu()`

**Comment by JarekDerp** [comment]
> can someone try `sigma = extra_options["sigmas"].cpu()`

I tried this:
```
        cond_or_uncond = extra_options["cond_or_uncond"]
        
        sigma = extra_options["sigmas"].cpu()
        print(sigma)
        
        sigma = extra_options["sigmas"][0] if 'sigmas' in extra_options else None
        sigma = sigma.item() if sigma is not None else 999999999.9
```

and i got this error when i run it:
```
!!! Exception during processing !!!
Traceback (most recent call last):
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\execution.py", line 151, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\execution.py", line 81, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\execution.py", line 74, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 773, in apply_ipadapter
    self.ipadapter = IPAdapter(
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\custom_nodes\ComfyUI_IPAdapter_plus\IPAdapterPlus.py", line 369, in __init__
    self.image_proj_model.load_state_dict(ipadapter_model["image_proj"])
  File "C:\StabilityMatrix-win-x64\Data\Packages\ComfyUI\venv\lib\site-packages\torch\nn\modules\module.py", line 2041, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for ImageProjModel:
        size mismatch for proj.weight: copying a param with shape torch.Size([3072, 1280]) from checkpoint, the shape in current model is torch.Size([3072, 1024]).

Prompt executed in 5.04 seconds
```

**Comment by cubiq** [comment]
that is not related to the sigma

**Comment by JarekDerp** [comment]
I haven't tested it thoroughly though. It's possible that running this if statement like it was before might not work even with the .cpu().

The current solution works fine so I'm not sure if it's worth changing it.

**Comment by cubiq** [comment]
would you mind trying with the previous statement and .cpu()?

**Comment by cubiq** [comment]
Something like this could work
```
        sigma = extra_options["sigmas"].detach().cpu().numpy()[0] if 'sigmas' in extra_options else None
        sigma = sigma if sigma is not None else 999999999.9
```

**Comment by MythicalChu** [comment]
> Something like this could work
> 
> ```
>         sigma = extra_options["sigmas"].detach().cpu().numpy()[0] if 'sigmas' in extra_options else None
>         sigma = sigma if sigma is not None else 999999999.9
> ```

I confirm that the above does work

**Comment by JarekDerp** [comment]
I can confirm that even this works from the original timestepping implementation:

```
sigma = extra_options["sigmas"].detach().cpu().numpy()[0] if 'sigmas' in extra_options else 999999999.9
```

Performance-wise, it's the same as this:
```
        sigma = extra_options["sigmas"][0] if 'sigmas' in extra_options else None
        sigma = sigma.item() if sigma else 999999999.9
```
I didn't notice any speed differences.

**Comment by cubiq** [comment]
will be included in the next major release

**Comment by cubiq** [comment]
hopefully fixed
